src/panel/direction/DirectionPanel.jsx

```diff
@@ -200,16 +221,19 @@ export default class DirectionPanel extends React.Component {
       return;
     }
 
-    // If origin field is empty, set it
-    // or, if destination field is empty, set it
-    if (persistentPointState.origin === null) {
+    // If destination field is empty and focused, set it
+    // else, if origin field is empty (focused or not), set it
+    // else, if destination field is empty (focused or not), set it
+    if (persistentPointState.destination === null && this.focusedField === 'destination') {
```

Comment:
This new `focusedField` state value is a bit odd. Ideally, we would like to avoid using fired events to propagate state.
And as far as I can see, it doesn't handle "blur" events.

Why not this solution:
```suggestion
    if (persistentPointState.destination === null && document.activeElement.id === 'itinerary_input_destination') {
```
It's still a bit hackish though.
The "React way" would probably be to use a Ref attached the input (they are already used in DirectionForm.jsx), or to propagate the state with a prop function.

Reply:
unfortunately, as soon as you click the map to set a marker, document.activeElement is the canvas and not the field :(
But if there's a React way, I'm interested!

Reply:
hmm, you're right.  
The best way I can think of would be to define a new prop `onInputFocus` (similar to `onChangePoint`) in DirectionInput.jsx

Reply:
Good job figuring a way to make it work, as that focus part is really tricky and already caused me much trouble :+1: 

Events aren't how to do that in React, as we already have all the pieces in place with the component hierarchy and refs. As a rule of thumb we should restrain custom events to cases where two pieces of the app can't communicate otherwise, namely React UI and the map view.

I think I would use a `onInputFocus` callback like @amatissart  proposed, store the focused field in state and just check it in `setDirectionPoint`.
And as usual, don't hesitate to do it in another PR so we can discuss this point specifically and merge the main feature without waiting.